[TextField] Add support for vertical content above input

laurkim opened the pull request:
### WHY are these changes introduced?

Resolves #5342.

The new approach with autocomplete for `Tags` has applied tags displayed inline in the TextField above the input as opposed to a container adjacent to the TextField.

### WHAT is this pull request doing?

**[Storybook URL](https://5d559397bae39100201eedc1-yxxzuvddpn.chromatic.com/)**
Adds a new optional `verticalContent` prop that, when true, will allow content to be displayed before the input in the TextField.
When `verticalContent` is added, `input` and `verticalContent` are both wrapped in a container `div` that is then rendered instead of just `input` in the TextField.

Also adds a new breakpoint for the inline content to set the `max-height` to `328px` on desktop and `140px` on mobile.

**Note**: for below gifs, I have the tags inside a `Stack` component with `spacing="tight"` so disregard any density changes in that regard, because Tags are normally in a `Stack` component with `spacing="extraTight"` in Admin.
    <details>
      <summary>TextField with vertical content that reaches max height — desktop</summary>
      <img src="https://user-images.githubusercontent.com/26749317/160918999-d69b63ec-ecfa-42b9-b2c6-6bd769517aaa.gif" alt="Textfield with vertical content that reaches max height — desktop">
    </details>
    <details>
      <summary>TextField with vertical content that doesn't reach max height — desktop</summary>
      <img src="https://user-images.githubusercontent.com/26749317/160919010-3e6116b6-70c4-4414-8fdc-8c7b325d19fe.gif" alt="Textfield with vertical content that doesn't reach max height — desktop">
    </details>
    <details>
      <summary>TextField with vertical content that reaches max height — mobile</summary>
      <img src="https://user-images.githubusercontent.com/26749317/160919008-a9ac4c70-19e7-4a2a-ba87-3103af44fbc3.gif" alt="Textfield with vertical content that reaches max height — mobile">
    </details>
    <details>
      <summary>TextField with vertical content that doesn't reach max height — mobile</summary>
      <img src="https://user-images.githubusercontent.com/26749317/160919012-6afdeb50-ef99-41bf-a6f2-1b89a2037b20.gif" alt="Textfield with vertical content that doesn't reach max height — mobile">
    </details>
    <details>
      <summary>New example for TextField on styleguide</summary>
      <img src="https://user-images.githubusercontent.com/26749317/160920985-3381d43d-3e55-4918-9177-39b0f1fb1ea6.png" alt="New example for TextField on styleguide">
    </details>

## <!-- ℹ️ Delete the following for small / trivial changes -->

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

<!--
  Give as much information as needed to experiment with the component
  in the playground.
-->

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React, {useState, useCallback, useEffect} from 'react';
import {CirclePlusMinor} from '@shopify/polaris-icons';

import {Page, Combobox, Listbox, TextStyle, Stack, Tag, Card} from '../src';

const tags = [
  'one',
  'two',
  'three',
  'four',
  'five',
  'six',
  'seven',
  'eight',
  'nine',
  'ten',
  'eleven',
  'twelve',
  'thirteen',
  'fourteen',
  'fifteen',
  'sixteen',
  'seventeen',
  'eighteen',
  'nineteen',
  'twenty',
  'twenty one',
  'twenty two',
  'twenty three',
  'twenty four',
  'twenty five',
  'twenty six',
  'twenty seven',
  'twenty eight',
  'twenty nine',
  'thirty',
  'thirty one',
  'thirty two',
  'thirty three',
  'thirty four',
  'thirty five',
  'thirty six',
  'thirty seven',
  'thirty eight',
  'thirty nine',
  'forty',
  'forty one',
  'forty two',
  'forty three',
  'forty four',
  'forty five',
  'tag one',
  'tag two',
  'tag three',
  'tag four',
  'tag five',
  'tag six',
  'tag seven',
  'tag eight',
  'tag nine',
  'tag ten',
  'tag eleven',
  'tag twelve',
  'tag thirteen',
  'tag fourteen',
  'tag fifteen',
  'tag sixteen',
  'tag seventeen',
  'tag eighteen',
  'tag nineteen',
  'tag twenty',
  'tag twenty one',
  'tag twenty two',
  'tag twenty three',
  'tag twenty four',
  'tag twenty five',
  'tag twenty six',
  'tag twenty seven',
  'tag twenty eight',
  'tag twenty nine',
  'tag thirty',
  'tag thirty one',
  'tag thirty two',
  'tag thirty three',
  'tag thirty four',
  'tag thirty five',
  'tag thirty six',
  'tag thirty seven',
  'tag thirty eight',
  'tag thirty nine',
  'tag forty',
  'tag forty one',
  'tag forty two',
  'tag forty three',
  'tag forty four',
  'tag forty five',
];

export function Playground() {
  const [inputValue, setInputValue] = useState<string>('');
  const [selectedTags, setSelectedTags] = useState(tags);
  const [filteredTags, setFilteredTags] = useState<string[]>(tags);

  const handleFilterOptions = useCallback(
    (inputValue: string) => {
      if (inputValue === '') {
        setFilteredTags(selectedTags);
        return;
      }

      const filterRegex = new RegExp(inputValue, 'i');
      const nextFilteredTags = tags.filter((option) =>
        option.match(filterRegex),
      );

      setFilteredTags(nextFilteredTags);
    },
    [selectedTags],
  );

  const handleInputChange = useCallback((value) => {
    setInputValue(value);
  }, []);

  useEffect(() => {
    handleFilterOptions(inputValue);
  }, [inputValue, handleFilterOptions]);

  const handleActiveOptionChange = useCallback((option) => {}, [inputValue]);

  const handleSelect = useCallback(
    (selected) => {
      if (selectedTags.includes(selected)) {
        setSelectedTags(selectedTags.filter((option) => option !== selected));
      } else {
        setSelectedTags([selected, ...selectedTags]);
        handleInputChange('');
      }
    },
    [selectedTags, handleInputChange],
  );

  const formatOption = useCallback(
    (option) => {
      if (!inputValue) return option;

      const matchIndex = option.toLowerCase().indexOf(inputValue);
      const start = option.slice(0, matchIndex);
      const highlight = option.slice(
        matchIndex,
        matchIndex + inputValue.length,
      );
      const end = option.slice(matchIndex + inputValue.length, option.length);

      return (
        <Listbox.TextOption selected={selectedTags.includes(option)}>
          <p>
            {start}
            <TextStyle variation="strong">{highlight}</TextStyle>
            {end}
          </p>
        </Listbox.TextOption>
      );
    },
    [inputValue],
  );

  const onRemove = (removedTag: string) => {
    const updatedTags = filteredTags.filter((tag) => tag !== removedTag);
    setFilteredTags(updatedTags);
    setSelectedTags(updatedTags);
  };

  const visibleTagMarkup = filteredTags.map((tag) => (
    <Tag onRemove={onRemove && (() => onRemove(tag))} key={`${tag}-tag`}>
      {tag}
    </Tag>
  ));

  const verticalContentMarkup = (
    <Stack spacing="tight" alignment="center">
      {visibleTagMarkup}
    </Stack>
  );

  const input = (
    <Combobox.TextField
      labelHidden
      disabled={false}
      label="Tags"
      value={inputValue}
      type="text"
      autoComplete="off"
      ariaAutocomplete="list"
      onChange={handleInputChange}
      verticalContent={verticalContentMarkup}
    />
  );

  const createActionMarkup = inputValue ? (
    <Listbox.Action
      accessibilityLabel={`Add: ${inputValue}`}
      value={inputValue}
      icon={CirclePlusMinor}
    >
      <p>
        <TextStyle variation="strong">Add: </TextStyle>
        {inputValue}
      </p>
    </Listbox.Action>
  ) : null;

  const optionMarkup =
    filteredTags.length > 0
      ? filteredTags.map((option) => {
          return (
            <Listbox.Option
              key={option}
              value={option}
              selected={selectedTags.includes(option)}
            >
              {formatOption(option)}
            </Listbox.Option>
          );
        })
      : null;

  const listboxMarkup =
    optionMarkup || createActionMarkup ? (
      <Listbox
        accessibilityLabel="Find or create tags"
        onSelect={handleSelect}
        onActiveOptionChange={handleActiveOptionChange}
        enableKeyboardControl
      >
        {optionMarkup}
        {createActionMarkup}
      </Listbox>
    ) : null;

  return (
    <Page title="Playground">
      <Card title="Tags" subdued>
        <Card.Section>
          <Combobox activator={input} allowMultiple>
            {listboxMarkup}
          </Combobox>
        </Card.Section>
      </Card>
    </Page>
  );
}
```

</details>

### 🎩 checklist

- [x] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
- [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
- [x] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
- [x] Updated the component's `README.md` with documentation changes
- [x] [Tophatted documentation](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting%20documentation.md) changes in the style guide
- [x] For visual design changes, ping @ sarahill to update the Polaris UI kit

<!--
  When tophatting, please check the UNRELEASED entry for consistency, as per the [guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md).
-->

laurkim committed:
[TextField] Add option to display content inline

Adds a new optional `verticalContent` prop that, when true, will allow content to be displayed before the input in the TextField.

When `verticalContent` is added, `input` and `verticalContent` are both wrapped in a container `div` that is then rendered instead of just `input` in the TextField.

chloerice commented on polaris-react/UNRELEASED.md:
```diff
@@ -7,6 +7,7 @@ Use [the changelog guidelines](/documentation/Versioning%20and%20changelog.md) t
 ### Enhancements
 
 - Change types for DataTable `totalsName` prop to allow for ReactNode ([#5454](https://github.com/Shopify/polaris/pull/5365/))
+- Added `verticalContent` prop to `TextField` to allow inline vertical content with `max-height` styling and examples ([#5392](https://github.com/Shopify/polaris/pull/5392/))
```
```suggestion
- Added a `verticalContent` prop to `TextField` to support rendering markup above the input value ([#5392](https://github.com/Shopify/polaris/pull/5392/))
```

chloerice commented on polaris-react/src/components/TextField/README.md:
```diff
@@ -482,6 +482,41 @@ function PrefixExample() {
 
+### Text field with vertical content
+
+Use to include custom vertical content before the input in the TextField.
```
```suggestion
Use to include custom content above the input value, like selected tags.
```

chloerice commented on polaris-react/src/components/TextField/TextField.tsx:
```diff
@@ -52,6 +52,8 @@ interface NonMutuallyExclusiveProps {
   prefix?: React.ReactNode;
   /** Text to display after value */
   suffix?: React.ReactNode;
+  /** Content to vertically display before value */
```
```suggestion
  /** Content to vertically display above the input value */
```

chloerice commented on polaris-react/src/components/TextField/TextField.tsx:
```diff
@@ -481,7 +500,7 @@ export function TextField({
         >
           {prefixMarkup}
-          {input}
+          {verticalContent ? inputWithVerticalContentMarkup : input}
```
Let's move the conditional out of the return by making it a variable.

```suggestion
const inputMarkup = verticalContent ? inputWithVerticalContentMarkup : input;
...
          {inputMarkup}
```

chloerice reviewed (approved):
🚀 🚀 🚀

laurkim committed:
Update UNRELEASED.md

laurkim committed:
Update Autocomplete multiselect example

laurkim committed:
[DataTable] Update readme example

The example with increased density and zebra striping was causing compiling errors in the styleguide when running `yarn run build-consumer polaris-styleguide`

chloerice committed:
Undo playground changes

chloerice committed:
Ensure contents of vertical content is full width
